openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -1480,6 +1487,49 @@ TCN_IMPLEMENT_CALL(void, SSLContext, setCertRequestedCallback)(TCN_STDARGS, jlon
     }
 }
 
+static int ssl_servername_cb(SSL *ssl, int *ad, void *arg)
+{
+    JNIEnv *e = NULL;
+    tcn_ssl_ctxt_t *c = arg;
+
+    const char *servername = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
+    if (servername != NULL) {
+        tcn_get_java_env(&e);
+        if((*e)->CallBooleanMethod(e, c->sni_hostname_matcher, c->sni_hostname_matcher_method, P2J(ssl), (*e)->NewStringUTF(e, servername)) == JNI_FALSE) {
+            return SSL_TLSEXT_ERR_ALERT_FATAL;
+        }
+    }
+    return SSL_TLSEXT_ERR_OK;
+}
+
+TCN_IMPLEMENT_CALL(void, SSLContext, setSniHostnameMatcher)(TCN_STDARGS, jlong ctx, jobject matcher)
+{
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+
+    UNREFERENCED(o);
+    TCN_ASSERT(ctx != 0);
+
+    if (matcher == NULL) {
+        SSL_CTX_set_tlsext_servername_callback(c->ctx, NULL);
+        SSL_CTX_set_tlsext_servername_arg(c->ctx, NULL);
+    } else {
+        jclass matcher_class = (*e)->GetObjectClass(e, matcher);
+        jmethodID method = (*e)->GetMethodID(e, matcher_class, "match", "(JLjava/lang/String;)Z");
+        if (method == NULL) {
+            return;
+        }
+        // Delete the reference to the previous specified matcher if needed.
+        if (c->sni_hostname_matcher != NULL) {
+            (*e)->DeleteLocalRef(e, c->sni_hostname_matcher);
+        }
+        c->sni_hostname_matcher = (*e)->NewGlobalRef(e, matcher);
```

Comment:
Don't know if it's applicable but maybe `NewWeakGlobalRef()` instead of `NewGlobalRef`?

```c
c->sni_hostname_matcher = (*e)->NewWeakGlobalRef(e, matcher);
```

```c
jobject matcher = (*e)->NewLocalRef(e, (jweak)c->sni_hostname_matcher);
if (!(*e)->IsSameObject(e, matcher, NULL)) {
  // Do something with the SNIMatcher
} else {
  // It appears the SNIMatcher got GC'd in Java land
}
(*e)->DeleteLocalRef(e, matcher);
```